probe/plugins/registry.go

```diff
@@ -0,0 +1,293 @@
+	)
+	// TODO: use of fs.Stat (which is syscall.Stat) here makes this linux specific.
+	if err := fs.Stat(path, &statT); err != nil {
+		return nil, err
+	}
+	switch statT.Mode & syscall.S_IFMT {
+	case syscall.S_IFDIR:
+		files, err := fs.ReadDir(path)
+		if err != nil {
+			return nil, err
+		}
+		for _, file := range files {
+			fpath := filepath.Join(path, file.Name())
+			s, err := r.sockets(fpath)
+			if err != nil {
+				log.Warningf("plugins: error loading path %s: %v", fpath, err)
+			}
+			result = append(result, s...)
+		}
+	case syscall.S_IFSOCK:
+		result = append(result, path)
+	}
+	return result, nil
+}
+
+// ForEach walks through all the plugins running f for each one.
+func (r *Registry) ForEach(f func(p *Plugin)) {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	paths := []string{}
+	for path := range r.pluginsBySocket {
+		paths = append(paths, path)
+	}
+	sort.Strings(paths)
+	for _, path := range paths {
+		f(r.pluginsBySocket[path])
+	}
+}
+
+// Implementers walks the available plugins fulfilling the given interface
+func (r *Registry) Implementers(iface string, f func(p *Plugin)) {
+	r.ForEach(func(p *Plugin) {
+		for _, piface := range p.Interfaces {
+			if piface == iface {
+				f(p)
+			}
+		}
+	})
+}
+
+// Name implements the Reporter interface
+func (r *Registry) Name() string { return "plugins" }
+
+// Report implements the Reporter interface
+func (r *Registry) Report() (report.Report, error) {
+	rpt := report.MakeReport()
+	r.Implementers("reporter", func(plugin *Plugin) {
+		pluginReport, err := plugin.Report()
+		if err != nil {
+			log.Errorf("plugins: error getting report from %s: %v", plugin.ID, err)
+			pluginReport = report.MakeReport()
+		}
+		pluginReport.Plugins = pluginReport.Plugins.Add(plugin.PluginSpec)
+		rpt = rpt.Merge(pluginReport)
+	})
+	return rpt, nil
+}
+
+// Close shuts down the registry. It can still be used after this, but will be
+// out of date.
+func (r *Registry) Close() {
+	r.cancel()
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	for _, plugin := range r.pluginsBySocket {
+		plugin.Close()
+	}
+}
+
+// Plugin is the implementation of a plugin. It is responsible for doing the
+// plugin handshake, gathering reports, etc.
+type Plugin struct {
+	xfer.PluginSpec
+	context context.Context
+	socket  string
+	client  *http.Client
+	cancel  context.CancelFunc
+	backoff backoff.Interface
+}
+
+// NewPlugin loads and initializes a new plugin. If client is nil,
+// http.DefaultClient will be used.
+func NewPlugin(ctx context.Context, socket string, client *http.Client, expectedAPIVersion string, handshakeMetadata map[string]string) *Plugin {
+	params := url.Values{}
+	for k, v := range handshakeMetadata {
+		params.Add(k, v)
+	}
+
+	ctx, cancel := context.WithCancel(ctx)
+	p := &Plugin{context: ctx, socket: socket, client: client, cancel: cancel}
+	f := p.handshake(ctx, expectedAPIVersion, params)
+	f() // try the first time synchronously
+	p.backoff = backoff.New(f, "plugin handshake")
```

Comment:
I think that the `initialbackoff` and `maxbackoff` should be smaller than the default values (i.e. imagine a plugin developer waiting for 1 minute before his plugin is read again after an error).